File: Sources/SetlistCore/Review.swift
import Foundation

/// One performance of one joke: what screen 2 compares.
public struct Take: Identifiable, Equatable, Sendable {
    public var id: UUID
    public var jokeID: UUID
    public var showID: UUID
    /// The longest laugh this take drew, in seconds — the take's headline number.
    public var peakLaughDuration: TimeInterval

    public init(id: UUID = UUID(), jokeID: UUID, showID: UUID, peakLaughDuration: TimeInterval) {
        self.id = id
        self.jokeID = jokeID
        self.showID = showID
        self.peakLaughDuration = peakLaughDuration
    }
}

/// The anomaly rule behind the mockup's "2.4σ below the rest" flag.
public enum TakeAnomaly {
    /// How far below the rest a take must fall to be flagged.
    public static let sigmaThreshold = 2.0
    /// Fewer comparison takes than this and there is no norm to deviate from.
    public static let minimumComparisonTakes = 3

    /// How many standard deviations below the other takes' mean this take sits.
    /// Nil when there are too few other takes, or when they are all identical
    /// and there is no spread to measure against.
    public static func deviation(of take: Take, amongOthers others: [Take]) -> Double? {
        guard others.count >= minimumComparisonTakes else { return nil }
        let values = others.map(\.peakLaughDuration)
        let mean = values.reduce(0, +) / Double(values.count)
        let variance = values.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Double(values.count - 1)
        let sigma = variance.squareRoot()
        guard sigma > 0 else { return nil }
        return (mean - take.peakLaughDuration) / sigma
    }

    public static func isAnomalous(_ take: Take, amongOthers others: [Take]) -> Bool {
        guard let deviation = deviation(of: take, amongOthers: others) else { return false }
        return deviation >= sigmaThreshold
    }
}

/// The countdown's colour. A helper, not the dominant element — it changes
/// colour rather than size (requirement 2.9).
public enum CountdownTone: String, Equatable, Sendable {
    case neutral
    case warning
    case over
}

public enum Countdown {
    /// Remaining time at which the countdown turns orange.
    public static let warningThreshold: TimeInterval = 180

    /// Remaining seconds; negative once the slot is over.
    public static func remaining(plannedLength: TimeInterval, elapsed: TimeInterval) -> TimeInterval {
        plannedLength - elapsed
    }

    public static func tone(remaining: TimeInterval) -> CountdownTone {
        if remaining <= 0 { return .over }
        if remaining <= warningThreshold { return .warning }
        return .neutral
    }
}

File: packages/setlist_core/lib/src/review.dart
import 'dart:math';

/// One performance of one joke: what the take-comparison screen compares.
class Take {
  Take({required this.jokeId, required this.showId, required this.peakLaughDuration});

  final String jokeId;
  final String showId;

  /// The longest laugh this take drew — the take's headline number.
  final Duration peakLaughDuration;
}

/// The anomaly rule behind the mockup's "2.4σ below the rest" flag.
abstract final class TakeAnomaly {
  /// How far below the rest a take must fall to be flagged.
  static const double sigmaThreshold = 2;

  /// Fewer comparison takes than this and there is no norm to deviate from.
  static const int minimumComparisonTakes = 3;

  /// How many standard deviations below the other takes' mean this take sits.
  ///
  /// Null when there are too few other takes, or when they are all identical
  /// and there is no spread to measure against — both are "unknown", and
  /// neither is a zero deviation.
  static double? deviation(Take take, {required List<Take> amongOthers}) {
    if (amongOthers.length < minimumComparisonTakes) return null;
    final values = [for (final t in amongOthers) t.peakLaughDuration.inMicroseconds / 1e6];
    final mean = values.reduce((a, b) => a + b) / values.length;
    var sumSquares = 0.0;
    for (final value in values) {
      sumSquares += (value - mean) * (value - mean);
    }
    // Sample standard deviation: these takes are the ones that happened, not
    // the population of every take the joke could ever draw.
    final sigma = sqrt(sumSquares / (values.length - 1));
    if (sigma <= 0) return null;
    return (mean - take.peakLaughDuration.inMicroseconds / 1e6) / sigma;
  }

  static bool isAnomalous(Take take, {required List<Take> amongOthers}) {
    final d = deviation(take, amongOthers: amongOthers);
    return d != null && d >= sigmaThreshold;
  }
}

/// The countdown's colour.
enum CountdownTone { neutral, warning, over }

abstract final class Countdown {
  /// Remaining time at which the countdown turns orange.
  static const Duration warningThreshold = Duration(minutes: 3);

  /// Remaining time; negative once the slot is over, because a comedian who is
  /// over needs to know by how much.
  static Duration remaining({required Duration plannedLength, required Duration elapsed}) =>
      plannedLength - elapsed;

  static CountdownTone tone(Duration remaining) {
    if (remaining <= Duration.zero) return CountdownTone.over;
    if (remaining <= warningThreshold) return CountdownTone.warning;
    return CountdownTone.neutral;
  }
}
